### PR TITLE
feat(tianmu): tianmu exporter support NULL type. (#1169)

### LINKED ISSUE #1169: feature: tianmu exporter support NULL type
**Is your feature request related to a problem? Please describe.**

1 user var "TIANMU_LOAD_NULL" to customize the file format;
2 enclosed char is not empyt, NULL type is empty without enclosed char;
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by lujiashun**
## Summary about this PR

Issue Number: close #1169

1 fix sql load when varchar column's char length(should be byte length) exceeds input-text byte length; 
2 support user varialbe "TIANMU_LOAD_NULL" to customize the export file content; 
3 fix the NULL value of exporter when enclosed char is not empty;

## Tests Check List

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/loader/parsing_strategy.cpp [hunk 1]
```diff
@@ -606,9 +606,9 @@ void ParsingStrategy::GetValue(const char *value_ptr, size_t value_size, ushort
 
   } else if (core::ATI::IsTxtType(ati.Type())) {
     // process escape characters
-    if (ati.CharLen() < (uint)value_size) {
+    if (ati.Precision() < (uint)value_size) {
```
Comment (RingsC) [inline storage/tianmu/loader/parsing_strategy.cpp]: 1: using c++ style type cast. 
2: how about the boundary checking, such as ati.Precision() == value_size.

File: storage/tianmu/exporter/data_exporter_txt.cpp [hunk 2]
```diff
@@ -92,14 +92,14 @@ void DEforTxt::PutRowEnd() {
     WriteChars(line_terminator_);
 }
 
-size_t DEforTxt::WriteString(const types::BString &str, int len, bool text_or_bin) {
+size_t DEforTxt::WriteString(const types::BString &str, int len, bool text_or_bin, bool isNull) {
```
Comment (hustjieke) [inline storage/tianmu/exporter/data_exporter_txt.cpp]: Change `isNull` to `is_null`.
```
The names of variables (including function parameters) and data members are all lowercase, with underscores between words. See:
```
https://google.github.io/styleguide/cppguide.html#Variable_Names

File: storage/tianmu/core/engine.h [hunk 3]
```diff
@@ -313,6 +313,7 @@ class ResultExportSender final : public ResultSender {
 };
 
 enum class tianmu_var_name {
+  TIANMU_TIMEOUT,
```
Comment (RingsC) [inline storage/tianmu/core/engine.h]: The first enum item should be set a default value.

File: storage/tianmu/exporter/data_exporter_txt.cpp [hunk 4]
```diff
@@ -92,14 +92,14 @@ void DEforTxt::PutRowEnd() {
...
   int res_len = 0;
   bool enclose_output{false};
 
   // If you omit the word OPTIONALLY, all fields are enclosed by the ENCLOSED BY character.
   // If you specify OPTIONALLY, the ENCLOSED BY character is used only to enclose values from columns that
   // have a string data type (such as CHAR, BINARY, TEXT, or ENUM):
-  if (str_qualifier_ && (text_or_bin || !opt_enclosed_)) {
+  if (str_qualifier_ && !isNull && (text_or_bin || !opt_enclosed_)) {
```
Comment (hustjieke) [inline storage/tianmu/exporter/data_exporter_txt.cpp]: The same problem as above.

File: storage/tianmu/exporter/data_exporter_txt.h [hunk 5]
```diff
@@ -46,9 +46,9 @@ class DEforTxt : public DataExporter {
   void WriteStringQualifier() { /*data_exporter_buf_->WriteIfNonzero(str_q);*/
   }
   void WriteDelimiter() { WriteChars(delimiter_); }
-  void WriteNull() { WriteString(nulls_str_.c_str(), (int)nulls_str_.length()); }
+  void WriteNull() { WriteString(nulls_str_.c_str(), (int)nulls_str_.length(), false, true); }
```
Comment (RingsC) [inline storage/tianmu/exporter/data_exporter_txt.h]: same as: c++ styple type cast

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/exporter/data_exporter_txt.cpp (hunk 4 above)
Comment (lujiashun) [reply storage/tianmu/exporter/data_exporter_txt.cpp]: OK, I will change it to follow Google code style;

File: storage/tianmu/exporter/data_exporter_txt.h (hunk 5 above)
Comment (lujiashun) [reply storage/tianmu/exporter/data_exporter_txt.h]: OK, I will fix it;

File: storage/tianmu/core/engine.h (hunk 3 above)
Comment (lujiashun) [reply storage/tianmu/core/engine.h]: OK, I will fix it;

File: storage/tianmu/exporter/data_exporter_txt.cpp (hunk 2 above)
Comment (lujiashun) [reply storage/tianmu/exporter/data_exporter_txt.cpp]: OK, I will fix it.

File: storage/tianmu/loader/parsing_strategy.cpp (hunk 1 above)
Comment (lujiashun) [reply storage/tianmu/loader/parsing_strategy.cpp]: 1. Ok, I will fix it.
2.  (ati.Precision() < (uint)value_size) means the input text exceeds the max length,  (ati.Precision() == value_size) is available

**Review by RingsC — APPROVED** [review]
nice, LGTM

**Review by duanfuxiang0 — APPROVED** [review]
LGTM

**Pull request merged**